### ISSUE TITLE
delete_dim does not work as expected

### ISSUE THREAD

**Issue by sunalbert** [clab/dynet#445]
Hi, When implementing dimension-wise mean, I find function: delete_dim in dim.h may not work as expected. 
If a tensor's dim is 
`Dim({2, 5, 7}, 10)`
and argument of delete_dim function `i` is 2, which means I wanna reduce a tensor with 4 rank(including batch dimension) along the third dimension. The expected result is Dim({2, 5}, 10), but in fact delete_dim 
returns Dim({2, 5, 7}, 10) , since the loop requires `i+1<nd`
Besides, in some scenarios, the batch dimension may also need to be reduced, but current delete_dim cannot handle it. Maybe we need a new delete_dim ,which can delete multi dimensions in one call and delete batch dimension when necessary.

**Comment by neubig** [comment]
There are two parts here. The first part of not working when `i` is 2 seems like a bug.

For the second part, deleting the batch dimension is as simple as `d.bd = 1`.

**Comment by sunalbert** [comment]
Thanks for your reply.I think a delete_dim function which can delete multi dimension will be better.In the following case:
the tensor's dim is Dim({2, 5, 7, 8}, 10), I want to  reduce the second(index is 1) and the third dimension(index is 2). The right result is Dim({2,8}, 10).
I have to call delete_dim twice, first calling delete_dim(1), resulting in Dim({2,7,8}, 10). Notice that the original third dimension has become the second dimension, it may result in bug, if user call delete_dim(2).
If we have a function delete_dim([1,2]),we can get right result simply.

**Comment by neubig** [comment]
If you think that's necessary, sure, feel free to add it (maybe with the name `delete_dims`)! `delete_dim` is an internal function that users won't be expected to use directly, so the interface is not super-important.

**Comment by sunalbert** [comment]
Hi, @neubig . I need some some suggestion from you about the new function: delete_dims and mean_reduce.
First, Let me explain why we need delete_dims. In some cases ,we need to reduce multi-dimensions.For example,in batch normalization, a tensor with rank 4({d[0], d[1], d[2]}, bd), we should get the average of it along d[0], d[1], and batch dimension.
In other frameworks, we can make it by calling mean_reduce(tensor, [0,1,3]). Since the batch dimension in dynet is supposed to be different from all others, I think a mean reduce interface implemented as following may be a choice:
```
mean_dim(tensor, dims, reduce_batch )
tensor: tensor
dims: a list of dimensions to be reduced
reduce_batch: bool value, whether to reduce along batch dimension
```
What do you think about it?

**Comment by neubig** [comment]
@sunalbert Can you give a reference (preferably equation in the paper, possibly example from another implementation) that explains this procedure in detail?

**Comment by sunalbert** [comment]
Ok.Usually, we think different channels represent diffetent features, so we want to normalize these features separately. The main idea is similar with traditional normalization. Batch normalization reducing dimensions except channel-dimension implemented by tensorflow as following:
```
beta = tf.get_variable(name='beta', shape=[n_out], initializer=tf.constant_initializer(0.0)
                               , trainable=True)
gamma = tf.get_variable(name='gamma', shape=[n_out], initializer=tf.random_normal_initializer(1.0, 0.02),
                                trainable=True)
batch_mean, batch_var = tf.nn.moments(x, [0, 1, 2], name='moments')  # data layout : [N H W C]
... other code
normed = tf.nn.batch_normalization(x, mean, var, beta, gamma...)
```

**Comment by sunalbert** [comment]
@neubig There is the [equation](https://sunalbert.github.io/img/in-post/post-batch-normailzation/algorithm.jpg) in the paper.To make it more concrete, I draw a simple [picture](https://sunalbert.github.io/img/in-post/post-batch-normailzation/bn_in_conv.png) to explain the details about it: all the elements included by the feature map with red border form the mini-batch B in equation.Finally we can get a average vector and it'size is channel_num.

**Comment by neubig** [comment]
OK, I was able to take a look at this.

Reading the original paper, it says in Section 3: "instead of whitening the features in layer inputs and outputs jointly we will normalize each scalar feature independently". This indicates to me, as I seemed to remember, that each feature is normalized independently, as opposed to doing some sort of dimension-by-dimension normalization.

To confirm, I also took a look at this implementation in Theano, and this seems to be confirming my suspicions:
https://github.com/shuuki4/Batch-Normalization/blob/master/BatchNormalization.py

If this is the case, this should be easy to implement in DyNet, as we would just have to take the average and variance over the minibatch.

**Comment by sunalbert** [comment]
There are two kinds of BN(mode==0  mode == 1) implemented in Theano, one is simple/normal BN, the other is global BN which is often used in Conv2D. As you mentioned, simple BN is easy to implement in Dynet. Do you think it is necessary to implement global BN in Dynet, after all now we have con2d function.

**Comment by neubig** [comment]
OK, that sounds reasonable! Thanks for the clarification @sunalbert

**Comment by sunalbert** [comment]
@neubig Ok. If there is no other things to be confirmed, I will start to implement batch norm and ohter related operations. These operations interfaces are designed as following:
```
delete_dims(std::vector dims, bool reduce_batch)
mean_dim(Tensor input, std::vector axis_to_be_reduced)
variance_dim(Tensor input, std::vector axis_to_be_reduced)
batch_norm() # need more discussion
```
Notice that batch_norm operation will be pending until we have an operation that can get average value of specified parameter, because we need moving_mean and moving_variance when performing batch norm in testing phase. So we can have detailed discussion about the interface of  batch_norm when it is completed.

**Comment by neubig** [comment]
Fixed by https://github.com/clab/dynet/pull/477
